README.md
# "Reverse Engineering for Beginners" free ebook

Topics discussed: x86/x64, ARM/ARM64, MIPS, Java/JVM.

Compiled versions can be found here: https://beginners.re/

See also the [ChangeLog](https://beginners.re/current-tree/ChangeLog) for the latest changes.

Want to contribute in some way? Read [here](https://beginners.re/current-tree/CONTRIBUTING.md).
Do not worry to bother me while writing me about any petty mistakes you found, even if you are not very confident.
I'm writing for beginners, after all, so beginners' opinions and comments are crucial for my job.

Want to translate this book into some other language? Read [here](https://beginners.re/current-tree/Translation.md).



ChangeLog
(This file is poorly maintained...)

12-Sep-2020: Some stat: 227k lines in all text files, 91k lines in EN/RU versions.
             2377 TeX files, 312 C/C++ files, 255 PNG files, 1600 assembly listings.

 9-Sep-2020: The size of English/Russian/French version has been increased from ~1000 pp. to ~1300-1400 pp.
             It's because now margins have normal size, as defined by default in TeX.
             Now it would be easier to read the book (both online and printed), because optimial length of a line is ~66 characters.
             ( https://en.wikipedia.org/wiki/Line_length )
             Yes, readers would need to print more, but on a plus side, the printed book would be easier to bind.

             Why I choose to have small margins initially? Primary reason is to fit all the (wide) listings.

28-Nov-2019: The book isn't hosted on GitHub anymore. For the source code, see: https://beginners.re/src/
   Nov-2019: Many blog posts from yurichev.com/blog has been copypasted into the book
   Jan-2019: There was a time when \IT macro was used for italic text and \ITA for Italian text.
             To be consistent, I'm using \emph for italic text now and \IT for Italian text.
             True for filename suffixes as well.

             Also, \JA macro is now used for Japanese, instead of \JPN.

   Jan-2019: syntax highlighting (comments only so far)
02-Jul-2018: DBI examples: minesweeper and WinRAR
15-Sep-2016: More fundamentals and examples
14-Sep-2016: More of my blog posts are copypasted into the book
06-Sep-2016: Blog posts about FAT12 and fortune file has been copypasted into the book
05-Sep-2016: Blog posts about entropy and encrypted DB case #1 has been
             copypasted into the book
22-Jul-2016: Russian only: new example: ancient "Marriage" DOS game hack
20-Jul-2016: The book is now licensed under CC BY-SA 4.0 (like Wikipedia)
             instead of CC NC-ND 3.0.
16-Jul-2016: Blog posts about C/C++ pointers has been copypasted into the book
14-Jul-2016: No more LITE version, which was unpopular
04-Nov-2015: All exercises has been moved to http://challenges.re
21-Mar-2015: toupper() function example
18-Mar-2015: Lite (introductory) version
01-Mar-2015: Java examples
04-Dec-2014: restructured: "code patterns" part splitted into "code patterns" and "more advanced examples"
19-Nov-2014: Norton Guide: simplest XOR encryption
11-Nov-2014: a lot of MIPS examples	
20-Oct-2014: more about TLS and TLS callbacks.
03-Oct-2014: atoi() example.
24-Sep-2014: calculating machine epsilon examples (IEEE 754); Duff's device example;
             pointer to function argument example; variadic functions example;
             endianness example.
07-Aug-2014: ~800 pages!
01-Aug-2014: new code pattern: conditional operator
31-Jul-2014: Minesweeper (Windows XP) cheat example
30-Jul-2014: Color Lines game practical joke; Oracle .MSB files example
29-Jul-2014: simple cheating in Millenium DOS game
21-Jul-2014: Oracle .SYM files example
11-Jun-2014: The book has now ~700 pages.
09-Jun-2014: Fibonacci example
28-Apr-2014: A5 version for e-book readers
06-Apr-2014: new exercise: data compressor with a bug
30-Mar-2014: simple patching examples
29-Mar-2014: simple example of cracking simple hash-function by Z3 SMT-solver
25-Mar-2014: more exercises; short GCC, GDB, MSVC, OllyDbg cheatsheets.
05-Mar-2014: text strings, unicode. 520 pages!
04-Mar-2014: Fahrenheit to Celsius converter
03-Mar-2014: examples: "10 PRINT" demo, EICAR test file
27-Feb-2014: working with float point numbers using SIMD in x64
26-Feb-2014: more exercises added
25-Feb-2014: some basic examples are extended to x86-64
10-Feb-2014: win16 code examples, IDA cheatsheet.
05-Feb-2014: Obfuscation
03-Feb-2014: Incorrectly disassembled code examples
02-Feb-2014: Title was changed to a shorter "Reverse Engineering for Beginners".
31-Jan-2014: 64-bit in 32-bit environment
28-Jan-2014: SEH3/4
25-Jan-2014: CRT (win32)
16-Jan-2014: OpenMP
10-Jan-2014: C++ STL std::map, std::set
             Getting divisor in division by multiplication cases
05-Jan-2014: C++ STL std::vector, INT3.
04-Jan-2014: C++ STL std::list.
29-Dec-2013: C++ ostream, C++ STL std::string.
12-Dec-2013: Itanium; Profile-guided optimization (Oracle RDBMS example).
09-Dec-2013: REP/REPE/REPNE prefixes, MOVSx, SCASx, STOSx, CMPSx instructions.
01-Dec-2013: LOCK prefix, critical sections.
29-Nov-2013: (examples) 1) dongle-protected software for SCO OpenServer; 2) another for MS-DOS
28-Nov-2013: (examples) dongle-protected software for Mac OS Classic and PowerPC. The book has now 300 pages!
05-Nov-2013: (exercises) vulnerable toy webserver.
01-Nov-2013: a lot of exercises added.
29-Oct-2013: Not a quick anymore, but still for beginners.
             Title was changed to "An introduction to reverse engineering for beginners".
20-Oct-2013: LD_PRELOAD
19-Oct-2013: PE file format.
13-Aug-2013: pointers, C++ references.
10-Aug-2013: ARM CPU, C99 restrict keyword.
03-Jul-2013: C++ objects, Oracle RDBMS examples.
15-Mar-2013: Initial published version, TeX form.
12-Mar-2013: First ever git commit, TeX form.

...

2010: The very first parts has been published at http://wiki.conus.info in Russian language.



patterns/03_printf/MIPS/printf8.O0_IT.s
$LC0:
	.ascii	"a=%d; b=%d; c=%d; d=%d; e=%d; f=%d; g=%d; h=%d\012\000"
main:
; prologo funzione:
	addiu	$sp,$sp,-56
	sw	$31,52($sp)
	sw	$fp,48($sp)
	move	$fp,$sp
	lui	$28,%hi(__gnu_local_gp)
	addiu	$28,$28,%lo(__gnu_local_gp)
	lui	$2,%hi($LC0)
	addiu	$2,$2,%lo($LC0)
; passa il 5° argomento nello stack:
	li	$3,4			# 0x4
	sw	$3,16($sp)
; passa il 6° argomento nello stack:
	li	$3,5			# 0x5
	sw	$3,20($sp)
; passa il 7° argomento nello stack:
	li	$3,6			# 0x6
	sw	$3,24($sp)
; passa l' 8° argomento nello stack:
	li	$3,7			# 0x7
	sw	$3,28($sp)
; passa il 9° argomento nello stack:
	li	$3,8			# 0x8
	sw	$3,32($sp)
; passa il  1° argomento in \$a0: 
	move	$4,$2
; passa il 2° argomento in \$a1:
	li	$5,1			# 0x1
; passa il 3° argomento in \$a2:
	li	$6,2			# 0x2
; passa il 4° argomento in \$a3:
	li	$7,3			# 0x3
; chiama printf():
	lw	$2,%call16(printf)($28)
	nop
	move	$25,$2
	jalr	$25
	nop
; epilogo funzione:
	lw	$28,40($fp)
; imposta il valore di ritorno a 0:
	move	$2,$0
	move	$sp,$fp
	lw	$31,52($sp)
	lw	$fp,48($sp)
	addiu	$sp,$sp,56
; return
	j	$31
	nop
